compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/calc/BinaryArithmeticNode.java

```diff
@@ -235,6 +235,165 @@ private static ReassociateMatch findReassociate(BinaryNode binary, NodePredicate
         return null;
     }
 
+    private static ReassociateMatch findReassociativeOperand(BinaryArithmeticNode<?> binary, NodePredicate criterion) {
+        NodePredicate binaryCriterion = new ReassociativePredicate(binary);
+        ValueNode forX = binary.getX();
+        ValueNode forY = binary.getY();
+        if (binaryCriterion.apply(forX) && findReassociate((BinaryNode) forX, criterion) != null) {
+            return ReassociateMatch.x;
+        }
+
+        if (binaryCriterion.apply(forY) && findReassociate((BinaryNode) forY, criterion) != null) {
+            return ReassociateMatch.y;
+        }
+        return null;
+    }
+
+    private static class ReassociativePredicate implements NodePredicate {
+        private final BinaryArithmeticNode<?> binary;
+
+        ReassociativePredicate(BinaryArithmeticNode<?> binary) {
+            this.binary = binary;
+        }
+
+        @Override
+        public boolean apply(Node node) {
+            if (binary instanceof AddNode || binary instanceof SubNode) {
+                return node instanceof AddNode || node instanceof SubNode;
+            }
+            return node.getClass() == binary.getClass();
+        }
+    }
+
+    private static class LoopPhiPredicate implements NodePredicate {
+        private final BinaryArithmeticNode<?> binary;
+
+        LoopPhiPredicate(BinaryArithmeticNode<?> binary) {
+            this.binary = binary;
+        }
+
+        @Override
+        public boolean apply(Node node) {
+            if (node instanceof ValuePhiNode) {
+                ValuePhiNode iv = (ValuePhiNode) node;
+                return iv.values().contains(binary);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Tries to expose values which satisfy the criterion as an operand. This could create more
+     * opportunities for other optimizations like constant-fold and loop-invariant elimination. For
+     * example with a constantness criterion: {@code (a * 2) * b => (a * b) * 2}
+     *
+     * This method accepts only {@linkplain BinaryOp#isAssociative() associative} operations such as
+     * +, -, *, &, | and ^
+     */
+    public static ValueNode reassociateUnmatchedValues(BinaryArithmeticNode<?> node, NodePredicate criterion, NodeView view, boolean skipLoopPhiValues) {
+        ValueNode forX = node.getX();
+        ValueNode forY = node.getY();
+        assert node.getOp(forX, forY).isAssociative();
+
+        // Skip re-associations for expressions like "i = i + var". This may influence the counted
+        // loop detection. For example, if the loop counter is re-associated from "i += (var + 1)"
+        // to "i = (i + var) + 1", the counted loop will not be detected as expected.
+        if (skipLoopPhiValues) {
+            LoopPhiPredicate phiCriterion = new LoopPhiPredicate(node);
+            if (phiCriterion.apply(forX) || phiCriterion.apply(forY)) {
```

Comment:
also I am not sure this is restrictive enough. Have you tried with derived induction variables as well and what happens for re-associations of mul nodes?

Reply:
Regarding to the counted loop detection, I think re-associations for `mul ` nodes might have no influence to it, since the loop trip node should be a base induction variable, isn't it?
And except for the counted loop, do you mean it might have any other influence for re-associations like `a = 5 * i * 6  -> a = 30 * i`?

Reply:
The main concern to check the derived induction variables is that we need to get the `LoopsData` and detect the counted loop each time when canonicalizaion which I think influences the compile time more.

Reply:
no, I think we can come back there if necessary later. I was thinking of a case where we have a none re-associated basic induction variable and then a derived one and if the derived one can be destroyed if it is subject to re-association, but maybe that cannot happen, I just wanted to think it through...